velox/functions/lib/TimeUtils.h

```diff
@@ -123,4 +126,204 @@ struct InitSessionTimezone {
+  }
+  if (unit == "month") {
+    return DateTimeUnit::kMonth;
+  }
+  if (unit == "quarter") {
+    return DateTimeUnit::kQuarter;
+  }
+  if (unit == "year") {
+    return DateTimeUnit::kYear;
+  }
+  if (allowAbbreviated) {
+    if (unit == "dd") {
+      return DateTimeUnit::kDay;
+    }
+    if (unit == "mon" || unit == "mm") {
+      return DateTimeUnit::kMonth;
+    }
+    if (unit == "yyyy" || unit == "yy") {
+      return DateTimeUnit::kYear;
+    }
+  }
+  if (throwIfInvalid) {
+    VELOX_UNSUPPORTED("Unsupported datetime unit: {}", unitString);
+  }
+  return std::nullopt;
+}
+
+// Adjust datetime based on unit.
+FOLLY_ALWAYS_INLINE void adjustDateTime(
+    std::tm& dateTime,
+    const DateTimeUnit& unit) {
+  switch (unit) {
+    case DateTimeUnit::kYear:
+      dateTime.tm_mon = 0;
+      dateTime.tm_yday = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kQuarter:
+      dateTime.tm_mon = dateTime.tm_mon / 3 * 3;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kMonth:
+      dateTime.tm_mday = 1;
+      dateTime.tm_hour = 0;
+      dateTime.tm_min = 0;
+      dateTime.tm_sec = 0;
+      break;
+    case DateTimeUnit::kWeek:
+      // Subtract the truncation.
+      dateTime.tm_mday -= dateTime.tm_wday == 0 ? 6 : dateTime.tm_wday - 1;
+      // Setting the day of the week to Monday.
+      dateTime.tm_wday = 1;
+
+      // If the adjusted day of the month falls in the previous month
+      // Move to the previous month.
+      if (dateTime.tm_mday < 1) {
+        dateTime.tm_mon -= 1;
+
+        // If the adjusted month falls in the previous year
+        // Set to December and Move to the previous year.
+        if (dateTime.tm_mon < 0) {
+          dateTime.tm_mon = 11;
+          dateTime.tm_year -= 1;
+        }
+
+        // Calculate the correct day of the month based on the number of days
+        // in the adjusted month.
+        static const int daysInMonth[] = {
+            31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+        int daysInPrevMonth = daysInMonth[dateTime.tm_mon];
+
+        // Adjust for leap year if February.
+        if (dateTime.tm_mon == 1 && (dateTime.tm_year + 1900) % 4 == 0 &&
+            ((dateTime.tm_year + 1900) % 100 != 0 ||
+             (dateTime.tm_year + 1900) % 400 == 0)) {
+          daysInPrevMonth = 29;
+        }
+        // Set to the correct day in the previous month.
+        dateTime.tm_mday += daysInPrevMonth;
+      }
+      dateTime.tm_hour = 0;
+      dateTime.tm_min = 0;
+      dateTime.tm_sec = 0;
+      break;
+    case DateTimeUnit::kDay:
+      dateTime.tm_hour = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kHour:
+      dateTime.tm_min = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kMinute:
+      dateTime.tm_sec = 0;
+      break;
+    default:
+      VELOX_UNREACHABLE();
+  }
+}
+
+/// For fixed interval like minute, hour and day,
+/// we can truncate date by a simple arithmetic expression:
+/// floor(seconds / intervalSeconds) * intervalSeconds.
```

Comment:
Please avoid repeating the code in comments. Comment suggestion:
> /// Returns timestamp with seconds adjusted to the nearest lower multiple of the
/// specified interval. If the given seconds is negative and not an exact
/// multiple of the interval, it adjusts further down.